Implement Square Roots to FiniteFields Category

Brian-Heckel opened the pull request:
<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->
Fixes #39688
Fixes #39798

Added the following methods to the finite field category
- `quadratic_non_residue()`
- `square_root()` and `sqrt()`
- `_tonelli()` and `_cipolla()`: the algorithms for finding a square root
- `is_square()` Moved `is_square()` and `sqrt()` methods from element.pyx to their proper categories.

Most profiling tests showed that Tonelli's algorithm outperforms Cipolla. However a profile with a Finite Field with order of a Cullen prime resulted in Cipolla outperforming Tonelli.

I'm running the meson build and currently cannot find the way to build the html documentation.
### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

Brian-Heckel committed:
Added Square Roots to Finite Field Category

Added the following methods to the finite field
category
- quadratic_non_residue
- square_root and sqrt
- _tonelli and _cipolla: the algorithm for finding
  a square root
- is_square
Moved is_square and sqrt methods from element.pyx
to their proper categories.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+                square_root = self._tonelli()
+            if all:
+                return (square_root, -square_root)
+            return square_root
```
```suggestion
        square_root = sqrt
```

Methods can be aliased like this.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            b = pow(X, (q+1)//2, f)
+            return b
+
+        def sqrt(self, all=False, algorithm='Tonelli'):
```
```suggestion
        def sqrt(self, all=False, algorithm='tonelli'):
```

The algorithm parameter in Sage is typically a lowercase string.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            - ``algorithm`` -- string (default: 'Tonelli'); The algorithm to use
+              among ``'Tonelli'``, ``'Cipolla'``. Most finite fields will have
+              Tonelli to run fast, but in the case where the order minus 1 is
+              highly divisible by 2 Cipolla's method could be faster.
```
```suggestion
            - ``algorithm`` -- string (default: 'tonelli'); the algorithm to use
              among ``'tonelli'``, ``'cipolla'``. Tonelli is typically faster but has
              a worse worst-case complexity than Cipolla. In particular, if the
              field cardinality minus 1 is highly divisible by 2 and has a large
              odd factor then Cipolla may perform better.
```

Typically the algorithm options in Sage are in lowercase.

Clearer wording, since elements and the field both have something called an order I think it's more clear to use cardinality here.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+              a square
+
+            - if ``all=True``, a 2-tuple of square roots; raises an error if the
+             element is not a square
```
```suggestion
            - if ``all=True``, a tuple of all distinct square roots. This tuple can have
              length 0, 1, or 2 depending on how many distinct square roots the
              element has.
```

The IntegerMod implementation returns an empty list when `all=True` and the element is not a square.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            """
+            order = self.parent().order()
+            if not self.is_square():
+                raise ValueError("Element is not a square")
```
Return an empty tuple if `all=True`, to align with IntegerMod implementation.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            elif algorithm == 'Cipolla':
+                square_root = self._cipolla()
+            else:
+                square_root = self._tonelli()
```
```suggestion
            elif algorithm == 'tonelli':
                square_root = self._tonelli()
            else:
                square_root = self._cipolla()
```

It's probably marginally faster to check the default algorithm first (only really matters if we add more square root algorithms later).

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            we use tonelli's method for all other cases in general.
+            """
+            order = self.parent().order()
+            if not self.is_square():
```
Since all elements are squares in characteristic 2, defer this check until after the characteristic 2 case.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            `p = 2` or divisible by 2 then we can compute the
+            square root by `a^(q/2)` if q is 3 modulo 4 then we
+            can compute the square root by `a^((q+1)/4)` otherwise
+            we use tonelli's method for all other cases in general.
```
```suggestion
            The algorithms used come from chapter 6 of [BS1996]_.
            Let `q = p^n` be the order of the finite field, let `a` be the finite field element
            that we wish to find the square root of.
            
            - If `p = 2` then `a` is always a square, and the square root of `\sqrt{a} = a^{(q / 2)}`.
            - If `q \equiv 3 \pmod{4}` then if `a` is a square `\sqrt{a} = {a^((q+1) / 4)}`
            - For all other cases we use the algorithm given by the ``algorithm`` parameter.
```

Wording, I need to check that this formatting is okay in the generated HTML.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            can compute the square root by `a^((q+1)/4)` otherwise
+            we use tonelli's method for all other cases in general.
+            """
+            order = self.parent().order()
```
Since this is a method on elements and elements also have an order, perhaps change this to `field_order` or `cardinality`.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            order = self.parent().order()
+            if not self.is_square():
+                raise ValueError("Element is not a square")
+            if order % 2 == 0:
```
Rather than reduce modulo 2, it should be faster to just check if the characteristic of the finite field is equal to 2. The cardinality of the field is even iff the characteristic is 2.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+                raise ValueError("Element is not a square")
+            g = self.parent().quadratic_non_residue()
+            odd_order = (q - 1).odd_part()
+            even_exp = Integer.valuation(q-1, 2)
```
```suggestion
            even_exp, odd_order = (q - ZZ.one()).val_unit(2)
```

In .py files `1` is a Python int but `q` is going to be a Sage `Integer`. Using `ZZ.one()` avoids coercion overhead in the subtraction.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+            odd_order = (q - 1).odd_part()
+            even_exp = Integer.valuation(q-1, 2)
+            e = 0
+            for i in range(2, even_exp+1):
```
I'm curious if using Sage's `IntegerRange` instead of the python `range` has any performance impact.

Brian-Heckel replied:
I tested the difference of `IntegerRange` compared to `range` and found no observable difference between the two.

vincentmacri reviewed (changes requested):
First pass, mostly style things and wording in the documentation.

Brian-Heckel committed:
Apply suggestions from code review

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

Brian-Heckel committed:
Update src/sage/categories/finite_fields.py

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

Brian-Heckel committed:
Added Suggested Changes From Vincent

Brian-Heckel committed:
Ran Lint Tests and Fixed Stuff

Brian-Heckel committed:
Update src/sage/categories/finite_fields.py

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+            return b
+
+        def sqrt(self, all:bool =False, algorithm: str ='tonelli'):
+            """
```
```suggestion
            r"""
```

Raw strings prevent issues with backslashes.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+            return b
+
+        def _cipolla(self):
+            """
```
```suggestion
            r"""
```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+            return is_square
+
+        def _tonelli(self):
+            """
```
```suggestion
            r"""
```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
     class ElementMethods:
-        pass
+        def is_square(self) -> bool:
+            """
```
```suggestion
            r"""
```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
 
+        @cached_method
+        def quadratic_non_residue(self):
+            """
```
```suggestion
            r"""
```

vincentmacri reviewed (changes requested):
Docstring fixes

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+            a square root element
+
+            OUTPUT:
+            - `True` if the element is a square `False` if not
```
```suggestion
            - ``True`` if the element is a square ``False`` if not
```

Two backticks for code, one for LaTeX

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+            OUTPUT:
+            - `True` if the element is a square `False` if not
+
+            EXAMPLES::
```
```suggestion
            EXAMPLES::

```

There needs to be a blank line after EXAMPLES

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+            - A square root of the element; raises an error
+              if the element is not a square
+
+            EXAMPLES::
```
```suggestion
            EXAMPLES::

```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+            - A square root of the element; raises an error
+              if the element is not a square
+
+            EXAMPLES::
```
```suggestion
            EXAMPLES::

```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+                - a non-square element of the finite field; raises an error if
+                  the finite field is of even order
+
+            EXAMPLES::
```
```suggestion
            EXAMPLES::

```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,222 @@ def _element_of_factored_order(self, F):
+              length 0, 1, or 2 depending on how many distinct square roots the
+              element has.
+
+            EXAMPLES::
```
```suggestion
            EXAMPLES::

```

vincentmacri reviewed (changes requested):
Docstring formatting

Brian-Heckel committed:
Documentation Fixes

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

Brian-Heckel committed:
Fixed Ruff Lints

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+            """
+            # if the order is an even power of two
+            # then every element is a square
+            if self.order() % 2 == 0:
```
```suggestion
            if self.characteristic() == 2:
```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+            if self.order() % 2 == 0:
+                raise ValueError("There are no non-squares in finite fields of even order")
+            # uniformly randomly select elements for a non-square
+            # with probability 1/2 for a non-square
```
The IntegerModRing implementation loops through all field elements rather than choosing them randomly. That might be preferable because finite fields have to be iterable but I don't think need to have a random_element method. It would also make the algorithm deterministic (assuming the iterable is).

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+            if self.is_zero():
+                return True
+            char = self.parent().characteristic()
+            if char == 2:
```
```suggestion
            if self.parent().characteristic() == 2:
```

No need to create a variable you're only using once.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+        def _tonelli(self):
+            r"""
+            Returns a square root of the element if it exists
+            using tonelli's algorithm
```
```suggestion
            using Tonelli's algorithm
```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+        def _cipolla(self):
+            r"""
+            Returns a square root of the element if it exists
+            using cipolla's algorithm, more suited if order - 1
```
```suggestion
            using Cipolla's algorithm, more suited if order - 1
```

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+
+            ALGORITHM:
+
+            The algorithms used come from chapter 6 of [BS1996]_.
```
```suggestion
            The algorithms used come from chapter 7 of [BS1996]_.
```

I don't know why I suggested to change this from chapter 8 to 6 earlier. Tonelli and Cipolla are definitely in chapter 7.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+            Returns a square root of the element if it exists
+            using cipolla's algorithm, more suited if order - 1
+            is highly divisible by 2
+
```
Cipolla's algorithm only works for odd characteristic, that should be mentioned in the docstring.

However since this is a private function I don't think error checking/handling for this is necessary.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+            r"""
+            Returns a square root of the element if it exists
+            using tonelli's algorithm
+
```
Tonelli's algorithm only works for odd characteristic, that should be mentioned in the docstring.

vincentmacri reviewed (changes requested):
Docstrings, cleanup.

vincentmacri commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,227 @@ def _element_of_factored_order(self, F):
+                ValueError: Element is not a square
+                sage: k.quadratic_non_residue().sqrt(all=True)
+                ()
+
```
Add an example (with an explanation) of a square root computation over a finite field where Cipolla's algorithm is faster (for example, the Cullen prime we tested).

vincentmacri commented on src/sage/categories/commutative_rings.py:
```diff
@@ -576,7 +576,184 @@ def _pseudo_fraction_field(self):
+            - if ``all=False``, a square root; raises an error if ``extend=False``
+              and ``self`` is not a square
+
+            - if ``all=True``, a list of all the square roots (empty if
```
Might as well change this function to return a tuple instead of a list since you're touching this code anyway.

Brian-Heckel replied:
This breaks some doctests in `src/sage/rings/function_field/element_rational.pyx` on line 354, changing the doctest to return a tuple instead didn't resolve the error since it has to do with "object is not iterable".

vincentmacri replied:
Okay, we can leave changing this to a tuple for a future PR then.

Brian-Heckel committed:
Review Changes

- Changed order mod 2 call to characteristic equal in is_square()
- Capitalization in docstrings
- Correct chapter in docstring for sqrt
-

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

Brian-Heckel committed:
Documentation and Iteration instead of Random

Fixed Latex documentation and error message
capitalization. Instead of picking randomly
for a quadratic non residue we now iterate
through the field until we find one, making
sqrt and _tonelli deterministic, _cipolla is
still random pick.

Brian-Heckel committed:
Removed comment on Random Picking

Brian-Heckel committed:
Merge branch 'develop' into FiniteField-sqrt

fchapoton commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,232 @@ def _element_of_factored_order(self, F):
+
+        def _cipolla(self):
+            r"""
+            Returns a square root of the element if it exists
```
First line should use imperative mode, so "Return" and not "Returns". Same elsewhere in the new methods.

fchapoton commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,232 @@ def _element_of_factored_order(self, F):
+            is highly divisible by 2. Only works for finite fields
+            of odd characteristic.
+
+            OUTPUT: A square root of the element; raises an error
```
wrong alignement here

vincentmacri replied:
I can't tell what's wrong with the alignment here. Can you explain?

Brian-Heckel replied:
I'm assuming there needs to be a newline after the `OUTPUT:`, like in `quadratic_non_residue()`

vincentmacri replied:
I think you're right. From the examples in https://doc.sagemath.org/html/en/developer/coding_basics.html#the-docstring-of-a-function-content there should be a newline after `OUTPUT:` if the description is multiple lines.

Brian-Heckel replied:
Fixed in the Documentation Fixes Commit

fchapoton commented on src/sage/categories/finite_fields.py:
```diff
@@ -243,5 +246,232 @@ def _element_of_factored_order(self, F):
+                square_root = self**exponent
+                if all:
+                    # we return a 1-tuple because the GF implementation does it
+                    return (square_root)
```
this is not a one-tuple, missing a comma

Brian-Heckel replied:
Fixed in the Documentation Fixes commit

vincentmacri commented on src/sage/categories/commutative_rings.py:
```diff
@@ -576,7 +576,184 @@ def _pseudo_fraction_field(self):
+
+            EXAMPLES::
+
+                sage: # needs sage.libs.pari
```
Are these `# needs` tags actually needed here? I don't think this test requires PARI, unless I'm missing something.

vincentmacri replied:
If it isn't needed, there are several other `# needs` tags that should also be removed.

vincentmacri replied:
Never mind, the `# needs` tag came from `element.pyx` so presumably it is needed.

vincentmacri commented on src/sage/categories/commutative_rings.py:
```diff
@@ -576,7 +576,184 @@ def _pseudo_fraction_field(self):
+                sage: R.<x> = QQ[]
+                sage: a = 2*(x+1)^2 / (2*(x-1)^2); a.sqrt()
+                (x + 1)/(x - 1)
+                sage: sqrtx=(1/x).sqrt(name='y'); sqrtx
```
```suggestion
                sage: sqrtx = (1/x).sqrt(name='y'); sqrtx
```

Brian-Heckel committed:
Documentation Fixes

Brian-Heckel committed:
Fixed equal spacing in doctest

Co-authored-by: Vincent Macri <vincent.macri@ucalgary.ca>

fchapoton commented on src/sage/categories/commutative_rings.py:
```diff
@@ -576,7 +576,184 @@ def _pseudo_fraction_field(self):
+            Return whether or not the ring element ``self`` is a square.
+
+            If the optional argument root is ``True``, then also return
+
```
what are these spurious empty lines ?

Brian-Heckel replied:
This function and `sqrt` are copies from the `element.pyx` versions, I can remove the empty lines if wanted

vincentmacri replied:
You can go ahead and remove the empty lines, since the code is being modified anyway might as well.

Brian-Heckel committed:
Renamed `quadratic_non_residue` to `quadratic_nonresidue`

vincentmacri reviewed (approved):
No further comments (does this make GitHub stop complaining about me requesting changes?)

Brian-Heckel committed:
Removed Redundant Spacing in Documentation

Brian-Heckel committed:
Merge branch 'develop' into FiniteField-sqrt
